Spigot-API-Patches/0243-Added-SoundBuilder.patch

```diff
@@ -0,0 +1,349 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Fruxz <cedricspitzer@outlook.de>
+Date: Fri, 18 Dec 2020 01:02:39 +0100
+Subject: [PATCH] Added SoundBuilder
+
+
+diff --git a/src/main/java/com/destroystokyo/paper/sound/SoundBuilder.java b/src/main/java/com/destroystokyo/paper/sound/SoundBuilder.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..803c51548d3ea74c6e6872908d751b42e0a420be
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/sound/SoundBuilder.java
+@@ -0,0 +1,337 @@
++package com.destroystokyo.paper.sound;
++
++import com.google.common.collect.Lists;
++import org.bukkit.Location;
++import org.bukkit.Sound;
++import org.bukkit.World;
++import org.bukkit.entity.Player;
++import org.bukkit.util.NumberConversions;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.Collection;
++import java.util.List;
++
++/**
++ * Helps prepare a sound to be sent to players.
++ */
++public class SoundBuilder {
++
++    private Sound sound;
++    private Collection<Player> receivers;
++    private Location location;
++    private Float volume = 1F;
++    private Float pitch = 1F;
++
++    public SoundBuilder(Sound sound) {
++        this.sound = sound;
++    }
++
++    public SoundBuilder(Sound sound, Float volume) {
++        this.sound = sound;
++        this.volume = volume;
++    }
++
++    public SoundBuilder(Sound sound, Float volume, Float pitch) {
++        this.sound = sound;
++        this.volume = volume;
++        this.pitch = pitch;
++    }
++
++    /**
++     * Plays the sound to all receiving players (or all). This method is safe to use
++     * Asynchronously
++     *
++     * @return a reference to this object.
++     */
++    public SoundBuilder play() {
++
++        if (this.location == null) {
++            throw new IllegalStateException("Please specify location for this sound");
++        }
++
++        if (receivers == null || receivers.isEmpty()) {
```

Comment:
Move this logic into -Server, then we can create a single Sound packet and reuse it for sending to everyone.

Doing it like this will create a packet per player.

Reply:
Sorry, I'm afraid I couldn't really fully comprehend that statement ):

Reply:
what aikar is suggesting is that the implementation of this logic should be in the Paper-Server project, because it would allow you to construct ne sound packet, and send that.
currently, you are just calling playsound for each player, that will create one packet per player, which just created unnecessary garbage objects.

alternatively, which is slightly out of scope for this PR I guess, we could have a playSound method that takes a list of receivers.

Reply:
I am sorry to inform you that I do not have the possibility to do this at the moment. My machine, with which I make these changes, is unfortunately not functional at the moment.

Reply:
its ok, it doesnt have to be tomorrow, just come back to this when you have the time :)

Reply:
Hey, my main machine is now working again to be able to make changes. But I have to admit that I don't know enough about the server-internal functions and structures to add the SoundBuilder to the Paper-Server project successfully and cleanly, so that this would be acceptable for me and for you. The question is, what can be done? Remove the receivers? Unfortunately I don't know what to do in this situation.